File: sdks/python/examples/abort_query.py
"""
Abort a long-running query using asyncio.Event.

The SDK watches the abort_event and kills the forge process when set().
An error message is yielded and the generator terminates.

Run: uv run python examples/abort_query.py
"""
import asyncio
import sys

from forgecode import query


async def main() -> None:
    abort_event = asyncio.Event()

    async def abort_after(seconds: float) -> None:
        await asyncio.sleep(seconds)
        print("[abort] Cancelling query...")
        abort_event.set()

    asyncio.create_task(abort_after(3))

    async for message in query(
        "Write a detailed 5000-word essay on the history of computing.",
        options={"abort_event": abort_event},
    ):
        match message.type:
            case "system":
                print(f"[session] {message.session_id}")
            case "assistant":
                sys.stdout.write(message.content)
            case "error":
                print(f"\n[error] {message.error}", file=sys.stderr)


if __name__ == "__main__":
    asyncio.run(main())

File: sdks/typescript/examples/abort-query.ts
/**
 * Abort a long-running query using AbortController.
 *
 * The SDK listens for the abort signal and kills the forge process
 * with SIGKILL. An error message is yielded and the generator terminates.
 *
 * Run: bun run examples/abort-query.ts
 */
import { query } from "../src";

const ac = new AbortController();

// Abort after 3 seconds
setTimeout(() => {
  console.log("[abort] Cancelling query...");
  ac.abort();
}, 3000);

try {
  for await (const message of query({
    prompt: "Write a detailed 5000-word essay on the history of computing.",
    options: {
      abortController: ac,
    },
  })) {
    switch (message.type) {
      case "system":
        console.log(`[session] ${message.session_id}`);
        break;
      case "assistant":
        process.stdout.write(message.content);
        break;
      case "error":
        console.error(`\n[error] ${message.error}`);
        break;
    }
  }
} catch (err) {
  console.error("Caught:", err);
}
